File: ruby/bob/bob.rb
class Bob
  RESPONSE = {
    yelling: 'Whoa, chill out!',
    asking: 'Sure.',
    silence: 'Fine. Be that way!',
    statement: 'Whatever.'
  }.freeze

  Yelling = ->(sentence) { sentence == sentence.upcase && sentence =~ /[A-Z]/ }
  Asking  = ->(sentence) { sentence.end_with?("?") }
  Silence = ->(sentence) { sentence.strip.empty? }

  def hey(sentence)
    RESPONSE[answer(sentence)]
  end

  private

    def answer(sentence)
      case sentence
      when Yelling
        :yelling
      when Asking
        :asking
      when Silence
        :silence
      else
        :statement
      end
    end
end

File: scala/bob/src/main/scala/Bob.scala
class Bob {
  def hey(statement: String): String = {
    if (statement.trim.size == 0)
      "Fine. Be that way!"
    else if (statement.toUpperCase == statement && statement.toLowerCase != statement)
      "Whoa, chill out!"
    else if (statement.last == '?')
      "Sure."
    else
      "Whatever."
  }
}
